Modify dset/attr builders based on sidecar JSON (#677)
hdmf-dev/hdmf: docs/source/sidecar.rst

.. _modifying_with_sidecar:

Modifying an HDMF File with a Sidecar JSON File
===============================================

> Users may want to update part of an HDMF file without rewriting the entire file.

Reviewer: ```suggestion
Users may want to update part of an HDMF file without rewriting the entire file.
```

I think it would be useful to elaborate a little bit on this to clarify the intent and scope of the sidecar file, i.e., this is for small updates and corrections only.

> Users may want to update part of an HDMF file without rewriting the entire file.

Reviewer: ```suggestion
Users may want to update part of an HDMF file without rewriting the entire file.
```

I think it would be useful to elaborate a little bit on this to clarify the intent and scope of the sidecar file, i.e., this is for small updates and corrections only.

To do so, HDMF supports the use of a "sidecar" JSON file that lives adjacent to the HDMF file on disk and
specifies modifications to the HDMF file. Only a limited set of modifications are supported; for example, users can

> delete a dataset or attribute but cannot create a new dataset or attribute.

Reviewer: ```suggestion
hide a dataset or attribute so that it will not be read by HDFM but cannot create a new dataset or attribute.
```

I think ``delete`` is misleading since we are not actually deleting any data from a file but the JSON file can only indicate that the dataset/attribute should be ignored on read (maybe ``hide`` or ``invalid`` would be more precise).

Does ``delete`` also apply to groups?

Author: Good point. I'll make the change. For now, I have not allowed hiding of groups because the use case is unclear. But it is technically not very different from hiding of datasets.

Reviewer: I think a main use-case for hiding groups would instances of a data_type, e.g., to hide a TimeSeries that for some reason contains bad data. If it's trivial, then I think allowing to hide groups is something we could allow, but if it adds a lot of complexity then I would hold off until a specific need arises.

> delete a dataset or attribute but cannot create a new dataset or attribute.

Reviewer: ```suggestion
hide a dataset or attribute so that it will not be read by HDFM but cannot create a new dataset or attribute.
```

I think ``delete`` is misleading since we are not actually deleting any data from a file but the JSON file can only indicate that the dataset/attribute should be ignored on read (maybe ``hide`` or ``invalid`` would be more precise).

Does ``delete`` also apply to groups?

When HDMF reads an HDMF file, if the corresponding sidecar JSON file exists, it is
automatically read and the modifications that it specifies are automatically applied.

.. note::

  This default behavior can be changed such that the corresponding sidecar JSON file is ignored when the HDMF file
  is read by passing ``load_sidecar=False`` to ``HDMFIO.read()`` on the ``HDMFIO`` object used to read the HDMF file.

Allowed modifications
---------------------

Only the following modifications to an HDMF file are supported in the sidecar JSON file:

- Replace the values of a dataset or attribute with a scalar or 1-D array
- Delete a dataset or attribute

.. note::

  Replacing the values of a dataset or attribute with a very large 1-D array using the sidecar JSON file may not
  be efficient and is discouraged. Users should instead consider rewriting the HDMF file with the
  updated values.

Specification for the sidecar JSON file
---------------------------------------

The sidecar JSON file can be validated using the ``sidecar.schema.json`` JSON schema file

> located at the root of the HDMF repository.

Reviewer: Are sidecar files automatically validated by the validator as well?

> located at the root of the HDMF repository.

Reviewer: Are sidecar files automatically validated by the validator as well?

                    4,
                    5
                ],
                "dtype": "int8"
            },
            {
                "type": "delete",
                "description": "delete foo1/foo_holder/my_sub_data/attr6",
                "object_id": "993fef27-680c-457a-af4d-b1d2725fcca9",
                "relative_path": "foo_holder/my_sub_data/attr6"
            }
        ],
        "schema_version": "0.1.0"
    }

Specification for operations
----------------------------

All operations are required to have the following keys:

- ``"type"``: The type of modification to perform. Only "replace" and "delete" are supported currently.
- ``"description"``: A description of the specified modification.
- ``"object_id"``: The object ID (UUID) of the data type that is closest in the file hierarchy to the
  field being modified.
- ``"relative_path"``: The relative path from the data type with the given object ID to the field being modified.

Operations can result in invalid files, i.e., files that do not conform to the specification. It is strongly
recommended that the file is validated against the schema after loading the sidecar JSON. In some cases, the

> file cannot be read because the file is invalid.

Reviewer: ```suggestion
.. warning:

    Modifying a file via a sidecar file can result in files that are no longer compliant with the format
    specification of the file. E.g., we may ``delete`` a required dataset via a sidecar operation, resulting
    in an invalid file that in the worst case, may longer be readable because required arguments are missing.  
    It is strongly recommended that the file is validated against the schema after loading the sidecar JSON.

```

> file cannot be read because the file is invalid.

Reviewer: ```suggestion
.. warning:

    Modifying a file via a sidecar file can result in files that are no longer compliant with the format
    specification of the file. E.g., we may ``delete`` a required dataset via a sidecar operation, resulting
    in an invalid file that in the worst case, may longer be readable because required arguments are missing.  
    It is strongly recommended that the file is validated against the schema after loading the sidecar JSON.

```
